text/XXX-unwind-FFI.md

```diff
@@ -0,0 +1,286 @@
+  to unwind (`panic`) without aborting.
+* If this annotation is used anywhere in the dependency tree, 
+  generation of the final (binary) product will fail
+  unless the panic strategy is `unwind`
+  and a non-default panic runtime is specified.
+* Stabilize the `#![panic_runtime]` annotation (from
+  [RFC #1513](less-unwinding)).
+* Provide an `unwind` runtime in the standard library
+  that guarantees compatibility with the native exception mechanism
+  provided by the compiler backend.
+* Provide a Cargo option under `[profile.foo]` to specify a `panic` runtime.
+
+[less-unwinding]: https://github.com/rust-lang/rfcs/blob/master/text/1513-less-unwinding.md
+
+# Motivation
+[motivation]: #motivation
+
+This will enable resolving
+[rust-lang/rust#58794](rust-ffi-issue)
+without breaking existing code.
+
+Currently, unwinding through an FFI boundary is always undefined behavior.
+We would like to make the behavior safe
+by aborting when the stack is unwound to an FFI boundary.
+However, there are existing Rust crates
+(notably, wrappers around the `libpng` and `libjpeg` C libraries)
+that rely on the current implementation's compatibility
+with the native exception handling mechanism in
+GCC, LLVM, and MSVC.
+
+This RFC will give crate authors the ability
+to initiate stack-unwinding
+that can propagate through other languages' stack frames,
+as well as tools for ensuring compatibility with those languages.
+
+[rust-ffi-issue]: https://github.com/rust-lang/rust/issues/58794
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Consider an `extern "C"` function that may `panic`:
+
+```rust
+extern "C" fn may_panic(i: i32) {
+    if i < 0 {
+        panic!("Oops, I should have used u32.");
+    }
+}
+```
+
+In a future (TBD) version of Rust,
+calling this function with an argument less than
+zero will cause the program to be aborted.
+This is the only way the compiler can ensure that the function is safe,
+because there is no way for it to know whether or not the calling code
+supports the same implementation of stack-unwinding used for Rust.
+
+As a concrete example,
+this function may be invoked from C code on Linux or BSD
+compiled without support for native stack-unwinding.
+The C runtime would lack the necessary metadata
+to properly propagate the unwinding operation,
+so it would be undefined behavior to let the runtime attempt
+to unwind the C stack frames.
+
+However, if the C code is compiled
+with support for the same stack-unwinding mechanism
+used to by the Rust code,
+unwinding across the FFI boundary is well-defined and safe,
+and in fact it can be a useful way to handle errors
+when working with certain C libraries, such as `libjpeg`.
+
+Thus, the `may_panic` function may be makred `#[unwind(allow)]`,
+which ensures that the unwinding operation will be propagated to the caller
+rather than aborting the process.
+This annotation will also prevent the compiler
+from marking the function as `noexcept`
```

Comment:
Rust does not have `noexcept`, LLVM does have `nounwind`, and that's what we use.

> nounwind: This function attribute indicates that the function never raises an exception. If the function does raise an exception, its runtime behavior is undefined.

Reply:
Ah. Sorry, I should have double checked. Thanks.
```suggestion
from marking the function as `nounwind`
```